doc/lsst.daf.butler/datastores.rst

```diff
@@ -0,0 +1,79 @@
+    The template uses data dimensions to do this.
+    Generally the default setting will be usable although it can be tuned per `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID.
+formatters
+    Mapping of `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID to a specific formatter class that understands the associated Python type and will serialize it to a file artifact.
+    The formatters section also supports the definitions of write recipes (bulk configurations that can be selected for specific formatters) and write parameters (parameters that control how the dataset is serialized; note it is required that all serialized artifacts be readable by a formatter without knowing which write parameters were used).
+constraints
+    Specify `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID that will be accepted or rejected by this datastore.
+composites
+    Controls whether composite datasets are disassembled by the datastore.
+    By default composites are not disassembled.
+    Disassembly can be controlled by `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID.
+
+.. _daf_butler-config-lookups:
+
+Name Matching
+^^^^^^^^^^^^^
+
+Templates, formatters, constraints, and composites all use a standard look up priority.
+The order is:
+
+#. If there is an ``instrument`` in the data ID the first look up will be for a key that matches ``instrument<INSTRUMENT_NAME>``.
+   If there is a match the items within that part of the hierarchy will be matched in preference to those at the top-level.
+#. The highest priority is then the `~lsst.daf.butler.DatasetType` name.
+#. If the `~lsst.daf.butler.DatasetType` corresponds to a component of a composite the composite name will then be checked.
+#. If there is still no match the dimensions will be used.
+   Dimensions are specified by the presence of a ``+`` as a separator.
+   For example ``instrument+physical_filter+visit`` would match any `~lsst.daf.butler.DatasetType` that uses those three dimensions.
+#. The final match is against the `~lsst.daf.butler.StorageClass` name.
+
+In-Memory Datastore
+===================
+
+The `~lsst.daf.butler.datastores.inMemoryDatastore.InMemoryDatastore` currently only supports the ``constraints`` field.
+This allows the datastore to accept specific dataset types.
+
+In the future more features will be added to allow some form of cache expiry.
+
+ChainedDatastore
+================
+
+This datastore enables multiple other datastores to be combined into one.
+The dataset will be sent to every datastore in the chain, and success is reported if any of the datastores accepts the dataset.
+When a dataset is retrieved each datastore is asked for the dataset in turn and the first match is sufficient.
+This allows an in-memory datastore to be combined with a file-based datastore to enable simple in-memory retrieval for a dataset that has been persisted to disk.
+
```

Comment:
Is this something that can be created on-the-fly, or is it something that you need to know you want ahead of time when you create a repository?

Reply:
You can turn a single datastore entry into a chained datastore without any problem. Things go wrong if you replace it with something that doesn't have any idea where the previous data are stored. In theory you could upload all the data locally to S3 and switch from a posix datastore to s3datastore without a problem. You can change a posix to a chained posix+in-memory or posix+s3+in-memory. For data that are only in the original posix datastore the data will come from that datastore. New writes will go to the old posix plus the new s3 say.

Reply:
I probably should also say somewhere that a key design of datastore is that once you have written it with a specific formatter, you can change your configuration to use different file templates or associate different formatters with that dataset type and you can still read the data. Only new writes will be affected.